### DISCUSSION TITLE
Activation code change every user refreshes on EmailActivator::class action page

### DISCUSSION THREAD

**Discussion by warcooft** [codeigniter4/shield#1133] [Q&A]
The activation code/token always changes when the user refreshes the `auth/a/show` page controller. **_Can the activation code sent to the user be the same as the first code?_** In case the user does not receive a token then the user has the option to send a new token via a special button/link.

Sometimes I face the following scenario: 
The activation code is sent by the system but the message is delayed 1-2 minutes to the destination email, so the user thinks the first attempt failed and tries again or refresh the page. then the second message containing the activation code is delayed while the first message has just been received. Then the user enters the first code after being submitted. It turns out the first code is invalid because it has been replaced with a second code which is still delayed. and so on

I hope you guys understand what I mean, I'm not good at English. Thanks

**Comment by kenjis** [comment]
Create your custom action as you like based on the `EmailActivator`.
See http://shield.codeigniter.com/references/authentication/auth_actions/

**Reply by warcooft** [reply]
OK, I'll try it. Thanks in advance

**Reply by grimpirate** [reply]
I had this issue with my TOTP app. Basically, the createIdentity method of the EmailActivator class deletes all prior instances of the identity. I solved this by checking for an existing identity before creating one. You can compare/contrast with the [TOTPActivator](https://github.com/grimpirate/halberd/blob/develop/src/Authentication/Actions/TOTPActivator.php) class for ideas.
